### PR TITLE
Adding CG support, moving analyze_deps to analyze job

### PR DESCRIPTION
**Pull request by KarishmaGhiya**
@Azure/azure-sdk-eng

### CODE REVIEW

**Comment by adxsdk6** [comment]
Can one of the admins verify this patch?

**Comment by adxsdk6** [comment]
Can one of the admins verify this patch?

File: .azure-pipelines/client.yml [hunk 1]
```diff
@@ -50,6 +46,35 @@ jobs:
     displayName: 'Publish Artifacts'
```
Comment (scbedd) [inline .azure-pipelines/client.yml]: It won't let leave a comment further up, the Copy task is moving the `dependencies.html` to the artifact staging directory. This no longer happens as the file ain't present anymore. Please adjust the task in the `analyze` job to write directly to the `Build.ArtifactsStagingDirectory`. Delete the `Move Dependencies Report to Build Artifacts` step above.

File: .azure-pipelines/client.yml [hunk 2]
```diff
@@ -50,6 +46,35 @@ jobs:
...
+  - script: |
+     pip install wheel Jinja2
+    displayName: 'Prep Environment'
+    
+  - task: PythonScript@0
+    displayName: 'Analyze dependencies'
+    enabled: true
+    inputs:
+      scriptPath: 'scripts/analyze_deps.py'
+      arguments: '--verbose --out "$(Build.SourcesDirectory)/dependencies.html"'
```
Comment (scbedd) [inline .azure-pipelines/client.yml]: This will no longer be present. Adjust to `Build.ArtifactStagingDirectory`

File: .azure-pipelines/client.yml [hunk 3]
```diff
@@ -124,9 +149,16 @@ jobs:
       summaryFileLocation: '$(Build.SourcesDirectory)/coverage.xml'
       reportDirectory: '$(Build.SourcesDirectory)/htmlcov'
 
+  - task: ms.vss-governance-buildtask.governance-build-task-component-detection.ComponentGovernanceComponentDetection@0
```
Comment (scbedd) [inline .azure-pipelines/client.yml]: Remove. This is taken care of in the `analyze` step.

File: .azure-pipelines/client.yml [hunk 4]
```diff
@@ -50,6 +46,35 @@ jobs:
...
+  
+  pool:
+    vmImage: 'vs2017-win2016' 
+    
+  steps:
+  
+  - task: UsePythonVersion@0
+    displayName: 'Use Python 3.6'
+    inputs:
+      versionSpec: 3.6
```
Comment (mikeharder) [inline .azure-pipelines/client.yml]: Please move to variable, and replace this hardcoded version number (and the earlier hardcoded version) with the variable.

File: .azure-pipelines/client.yml (hunk 3 above)
Comment (mikeharder) [inline .azure-pipelines/client.yml]: Why are we running CG from two places in the pipeline?

File: .azure-pipelines/client.yml [hunk 5]
```diff
@@ -124,9 +149,16 @@ jobs:
...
+  - task: ms.vss-governance-buildtask.governance-build-task-component-detection.ComponentGovernanceComponentDetection@0
+    displayName: 'Component Detection'
+    condition: and(eq(variables['os.name'],'Windows'),succeeded())
+
 - job: Test_Alpha_Python
   timeoutInMinutes: 90
   condition: eq(variables['long_running_tests'], True)
+
+
+- job: Special_Python_Distro_Tests
```
Comment (mikeharder) [inline .azure-pipelines/client.yml]: Is this related to the CG change?

File: .azure-pipelines/client.yml [hunk 6]
```diff
@@ -2,6 +2,8 @@
 #  'build_targeting_string' comma separated glob strings selecting the packages that should be built. EG: "azure-keyvault,azure-mgmt-batch" or "azure-*"
 #  'codecov-python-repository-token' Is the codecov.io repository token.
 #  'long_running_tests' Is a boolean indicating that the long-running tests should be executed. 
+#  'EnableDetectorPip': true . Required for Component Governance Support
+
```
Comment (scbedd) [inline .azure-pipelines/client.yml]: Extra whitespace. Let's keep a max of 1 blank line to separate stuff.

File: .azure-pipelines/client.yml [hunk 7]
```diff
@@ -2,6 +2,8 @@
 #  'build_targeting_string' comma separated glob strings selecting the packages that should be built. EG: "azure-keyvault,azure-mgmt-batch" or "azure-*"
 #  'codecov-python-repository-token' Is the codecov.io repository token.
 #  'long_running_tests' Is a boolean indicating that the long-running tests should be executed. 
+#  'EnableDetectorPip': true . Required for Component Governance Support
```
[the remark this answers is not in the document]
Comment (KarishmaGhiya) [reply .azure-pipelines/client.yml]: This is required for the Component Governance Support

File: .azure-pipelines/client.yml (hunk 5 above)
Comment (KarishmaGhiya) [reply .azure-pipelines/client.yml]: No, this job is not related to CG support

File: .azure-pipelines/client.yml (hunk 3 above)
Comment (KarishmaGhiya) [reply .azure-pipelines/client.yml]: No, I forgot to remove that code from the Test job. We only need to run it once under Analyze job.

File: .azure-pipelines/client.yml (hunk 7 above)
[the remark this answers is not in the document]
Comment (mikeharder) [reply .azure-pipelines/client.yml]: Do you know more about this variable?  Why does CG require it to be set?  What happens if it's not set?  Why doesn't CG just default this internally to "true", instead of requiring everyone to set it?

File: .azure-pipelines/client.yml [hunk 8]
```diff
@@ -2,32 +2,27 @@
...
 
 trigger:
 - master
+variables:
+  python.version: '3.6'
```
Comment (mikeharder) [inline .azure-pipelines/client.yml]: Per the [guidelines](https://dev.azure.com/azure-sdk/internal/_wiki/wikis/internal.wiki?wikiVersion=GBwikiMaster&pagePath=%2FAzure%20DevOps%20Pipeline%20Guidance&pageId=2&anchor=variables), we decided to standardize on `CamelCase` for variables.  So please rename this from `python.version` to `PythonVersion`.  Existing pipelines will eventually be updated to match this standard.

File: .azure-pipelines/client.yml [hunk 9]
```diff
@@ -38,18 +33,42 @@ jobs:
...
+      scriptPath: 'scripts/analyze_deps.py'
+      arguments: '--verbose --out "$(Build.ArtifactStagingDirectory)/dependencies.html"'
+
+  - task: ms.vss-governance-buildtask.governance-build-task-component-detection.ComponentGovernanceComponentDetection@0
+    # ComponentGovernance is currently unable to run on pull requests of public projects. Running on non-PR
+    # builds should be sufficient.
+    condition: and(always(), ne(variables['System.TeamProject'], 'public'))
+    displayName: 'Component Detection'
 
   - task: PublishBuildArtifacts@1
```
Comment (mikeharder) [inline .azure-pipelines/client.yml]: I think `PublishBuildArtifacts` should be removed from the `Analyze` job.  It should only be needed in the `Build` job.

File: .azure-pipelines/client.yml (hunk 5 above)
Comment (mikeharder) [reply .azure-pipelines/client.yml]: Why is this change part of this PR?  If it's unrelated to CG support it should ideally be in a separate PR.

Comment (scbedd) [inline .azure-pipelines/client.yml]: In the Test stage, you still have the component governance present with `only windows` condition. Please remove it as it's already part of the build in the `analyze` step.

Comment (KarishmaGhiya) [reply .azure-pipelines/client.yml]: The job Special_Python_Distro_Tests is from an old commit and is no longer there in the master. This happened because the merge went wrong while resolving conflicts. But I pushed a new update, that has the correct changes that should be there.

File: .azure-pipelines/client.yml (hunk 9 above)
Comment (KarishmaGhiya) [reply .azure-pipelines/client.yml]: Actually the task 'Analyze Dependencies' is generating a file dependencies.html which we want to make a part of the artifacts. So effectively, this particular Publish Artifact step is actually publishing that file from the artifacts staging directory

Comment (mikeharder) [reply .azure-pipelines/client.yml]: Please add a comment explaining this

File: .azure-pipelines/client.yml (hunk 7 above)
[the remark this answers is not in the document]
Comment (mikeharder) [reply .azure-pipelines/client.yml]: If we've verified this is required for CG support, I'm fine merging this PR as-is, but we should followup with the CG team.

File: .azure-pipelines/client.yml [hunk 10]
```diff
@@ -6,28 +6,25 @@
 trigger:
 - master
 
+variables:
+  PythonVersion: '3.6'
+
 jobs:
 - job: 'Build_Target_Packages'
```
Comment (weshaggard) [inline .azure-pipelines/client.yml]: Can you please update this job to call it "Build" as we added in the guidelines?

File: .azure-pipelines/client.yml (hunk 9 above)
Comment (weshaggard) [reply .azure-pipelines/client.yml]: @bsiegel should we consider putting the dependency output into an different artifact zip file?

File: .azure-pipelines/client.yml [hunk 11]
```diff
@@ -83,6 +109,7 @@ jobs:
         os.name: 'MacOS'
```
Comment (weshaggard) [inline .azure-pipelines/client.yml]: Sorry if we are hijacking your PR but if you are trying to make this file match the guidelines can you also update these matrix strategy variables?

Comment (KarishmaGhiya) [reply .azure-pipelines/client.yml]: Haha, yes I just pushed a commit for the same

File: .azure-pipelines/client.yml (hunk 9 above)
Comment (bsiegel) [reply .azure-pipelines/client.yml]: So far, I have just been using directories under the default "drop" artifact, for example in Java we now have :

```
drop/
  java-8/
  java-lts/
  reports/
```

We could potentially have multiple named artifacts instead. I don't have a strong opinion but we would have to change a number of release pipelines

Comment (weshaggard) [reply .azure-pipelines/client.yml]: Oh no is that what the java artifacts look like now? The release pipelines currently try to publish everything in the artifacts. We should at least consider having a "packages" or "release" artifact that will only contain what we need to release to our package managers.

File: .azure-pipelines/client.yml [hunk 12]
```diff
@@ -37,60 +34,90 @@ jobs:
...
-    - 'Build_Target_Packages'
+    - 'Build'
+
   strategy:
     matrix:
       Linux_Python27:
-        os.name: 'Linux'
-        os.vmImage: 'ubuntu-16.04'
-        python.version: '2.7'
+        OSName: 'Linux'
```
Comment (weshaggard) [inline .azure-pipelines/client.yml]: @scbedd it doesn't even look like we use the OSName anywhere so we should probably just delete it. @mikeharder I see you do use OsName as part of a task description in the JS pipeline. Given that the strategy name is used in the job name I personally don't feel it is worth plumbing this extra information through. What do others think?

Comment (scbedd) [reply .azure-pipelines/client.yml]: Yeah. everywhere in the template we use os.name.

Comment (weshaggard) [reply .azure-pipelines/client.yml]: I must be missing the usages then as I don't see "os.name" or "OSName" being used anywhere in this pipeline.

Comment (scbedd) [reply .azure-pipelines/client.yml]: You are correct. It's been updated.
